src/tests/units/html_tests.cpp

```diff
@@ -0,0 +1,493 @@
+  }};
+  // clang-format on
+
+  {
+    std::string sentence_str("I like to drive this car.");
+    std::vector<string_view> sentence{
+        string_view(sentence_str.data() + 0, 1),   // 0.0 "I"
+        string_view(sentence_str.data() + 1, 5),   // 0.1 " like"
+        string_view(sentence_str.data() + 6, 3),   // 0.2 " to"
+        string_view(sentence_str.data() + 9, 6),   // 0.3 " drive"
+        string_view(sentence_str.data() + 15, 5),  // 0.4 " this"
+        string_view(sentence_str.data() + 20, 4),  // 0.5 " car"
+        string_view(sentence_str.data() + 24, 1),  // 0.6 "."
+        string_view(sentence_str.data() + 25, 0),  // 0.7 ""
+    };
+    response.source.appendSentence("", sentence.begin(), sentence.end());
+    response.source.appendEndingWhitespace("\n");
+  }
+
+  {
+    std::string sentence_str("Ich fahre gerne dieses Auto.");
+    std::vector<string_view> sentence{
+        string_view(sentence_str.data() + 0, 3),   // 0.0 "Ich"
+        string_view(sentence_str.data() + 3, 1),   // 0.1 " "
+        string_view(sentence_str.data() + 4, 4),   // 0.2 "fahr"
+        string_view(sentence_str.data() + 8, 1),   // 0.3 "e"
+        string_view(sentence_str.data() + 9, 6),   // 0.4 " gerne"
+        string_view(sentence_str.data() + 15, 7),  // 0.5 " dieses"
+        string_view(sentence_str.data() + 22, 5),  // 0.6 " Auto"
+        string_view(sentence_str.data() + 27, 1),  // 0.7 "."
+        string_view(sentence_str.data() + 28, 0),  // 0.8 ""
+    };
+    response.target.appendSentence("", sentence.begin(), sentence.end());
+    response.target.appendEndingWhitespace("\n");
+  }
+
+  html.Restore(response);
+
+  {
+    AnnotatedText source;
+    std::string sentence_str("<p>I <b>like</b> to <u>drive</u> this car.");
+    std::vector<string_view> sentence{
+        string_view(sentence_str.data() + 0, 4),   // 0.0 "<p>I"
+        string_view(sentence_str.data() + 4, 8),   // 0.1 " <b>like"
+        string_view(sentence_str.data() + 12, 7),  // 0.2 "</b> to"
+        string_view(sentence_str.data() + 19, 9),  // 0.3 " <u>drive"
+        string_view(sentence_str.data() + 28, 9),  // 0.4 "</u> this"
+        string_view(sentence_str.data() + 37, 4),  // 0.5 " car"
+        string_view(sentence_str.data() + 41, 1),  // 0.6 "."
+        string_view(sentence_str.data() + 42, 0),  // 0.7 ""
+    };
+    source.appendSentence("", sentence.begin(), sentence.end());
+    source.appendEndingWhitespace("</p>\n");
+
+    CHECK(AsTokens(response.source) == AsTokens(source));
+  }
+
+  {
+    AnnotatedText target;
+    std::string sentence_str("<p>Ich <u>fahre</u> <b>gerne</b> dieses Auto.");
+    std::vector<string_view> sentence{
+        string_view(sentence_str.data() + 0, 6),    // 0.0 "<p>Ich"
+        string_view(sentence_str.data() + 6, 4),    // 0.1 " <u>"
+        string_view(sentence_str.data() + 10, 4),   // 0.2 "fahr"
+        string_view(sentence_str.data() + 14, 1),   // 0.3 "e"
```

Comment:
This is a different example, the `<b></b>` here isn't empty and isn't lost, it just moved with the translation of `_like`, `_gerne`. Previously because the the empty space in 0.1 (now `"_<u>"`) was aligned with the `_like` source token, it would get the `p,b` taint from that token. But now I smear the taint of the most prevalent token of a word across the whole word (including that preceding space) and it gets the `p,u` taint from `_drive`.

The disappearing `<b></b>` is tested in a bit higher up, I'll mark it with a comment for you.

Reply:
taint lol